File: crates/plugins/src/terminal.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

//! Activation-owned PTYs with call-bound control and replayable output.

pub use crate::process::Error;
use crate::{call::Scope, process::Command};
use futures_util::future::BoxFuture;
pub use maka_runtime::{shell_run::ShellOutcome as Outcome, terminal::TerminalSize as Size};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Spawn {
    pub command: Command,
    pub size: Size,
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Control {
    #[serde(default)]
    pub text: String,
    pub size: Option<Size>,
}
#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[serde(transparent)]
pub struct Id(pub String);
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Written {
    pub accepted_bytes: usize,
    pub resized: bool,
}
#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Output {
    Data {
        sequence: u64,
        text: String,
    },
    Reset {
        sequence: u64,
        text: String,
        size: Size,
    },
    Closed,
}
pub struct Handle {
    pub id: Id,
    pub io: Arc<dyn Terminal>,
}
pub trait Terminals: Send + Sync {
    fn spawn(&self, call: Scope, input: Spawn) -> BoxFuture<'_, Result<Handle, Error>>;
    /// Rebind an activation-local lookup key; each operation rechecks permission.
    fn open(&self, call: Scope, id: Id) -> Result<Handle, Error>;
    /// Owner cleanup does not require a still-live calling scope.
    fn close(&self, id: Id) -> BoxFuture<'_, Result<(), Error>>;
}
/// Dropping this view does not end an instance-owned terminal. The activation
/// owns the process; this view only fixes the authority used to interact with it.
pub trait Terminal: Send + Sync {
    fn control(&self, input: Control) -> BoxFuture<'_, Result<Written, Error>>;
    fn next(&self) -> BoxFuture<'_, Result<Output, Error>>;
    fn wait(&self) -> BoxFuture<'_, Result<Outcome, Error>>;
    fn close(&self) -> BoxFuture<'_, Result<(), Error>>;
}

File: packages/plugin-sdk/src/terminal.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import type { ProcessCommand } from './process.js';

export interface TerminalSize {
  cols: number;
  rows: number;
}
export type TerminalOutput =
  | { kind: 'data'; sequence: number; text: string }
  /** Initial/recovery cut of the bounded raw output tail; replace prior output. */
  | { kind: 'reset'; sequence: number; text: string; size: TerminalSize }
  | { kind: 'closed' };
export type TerminalOutcome =
  | { kind: 'completed' }
  | { kind: 'exited'; code: number; message: string | null }
  | { kind: 'failed' | 'orphaned'; message: string }
  | { kind: 'timed_out' | 'cancelled'; message: string | null };
export interface TerminalWrite {
  /** Transport-accepted UTF-8 prefix, not application consumption. */
  acceptedBytes: number;
  resized: boolean;
}
export interface Terminal {
  readonly id: string;
  /** Raw UTF-8/ESC/C0 keystrokes (Enter is CR). Resize and input are serialized together. */
  write(text: string, size?: TerminalSize): Promise<TerminalWrite>;
  resize(size: TerminalSize): Promise<TerminalWrite>;
  /** One pending read; slow consumers receive an explicit reset, never silent truncation. */
  next(): Promise<TerminalOutput>;
  /** Wait for durable exit AND native resource cleanup, not merely output EOF. */
  wait(): Promise<TerminalOutcome>;
  /** Signal independently of input/output backpressure and confirm cleanup. */
  close(): Promise<void>;
}
export interface Terminals {
  /** Uses the same authorized OS sandbox as processes, plus Host PTY limits. */
  spawn(command: ProcessCommand, size?: TerminalSize): Promise<Terminal>;
  /** Rebind an instance terminal to a later invocation of the same Session. */
  open(id: string): Terminal;
}
